pkg/cri/server/sandbox_stats_windows.go

```diff
@@ -19,17 +19,397 @@ package server
+	containerstore "github.com/containerd/containerd/pkg/cri/store/container"
 	sandboxstore "github.com/containerd/containerd/pkg/cri/store/sandbox"
+	"github.com/containerd/containerd/pkg/cri/store/stats"
+	"github.com/containerd/typeurl"
+	runtime "k8s.io/cri-api/pkg/apis/runtime/v1"
 )
 
-func (c *criService) podSandboxStats(ctx context.Context, sandbox sandboxstore.Sandbox, stats interface{}) (*runtime.PodSandboxStats, error) {
-	return nil, fmt.Errorf("pod sandbox stats not implemented on windows: %w", errdefs.ErrNotImplemented)
+func (c *criService) podSandboxStats(
+	ctx context.Context,
+	sandbox sandboxstore.Sandbox) (*runtime.PodSandboxStats, error) {
+	meta := sandbox.Metadata
+
+	if sandbox.Status.Get().State != sandboxstore.StateReady {
+		return nil, fmt.Errorf("failed to get pod sandbox stats since sandbox container %q is not in ready state", meta.ID)
+	}
+
+	timestamp := time.Now()
+	podSandboxStats := &runtime.PodSandboxStats{}
+	podSandboxStats.Windows = &runtime.WindowsPodSandboxStats{}
+	podSandboxStats.Attributes = &runtime.PodSandboxAttributes{
+		Id:          meta.ID,
+		Metadata:    meta.Config.GetMetadata(),
+		Labels:      meta.Config.GetLabels(),
+		Annotations: meta.Config.GetAnnotations(),
+	}
+
+	metrics, containers, err := c.listWindowsMetricsForSandbox(ctx, sandbox)
+	if err != nil {
+		return nil, fmt.Errorf("failed to obtain container stats during podSandboxStats call: %w", err)
+	}
+
+	statsMap, err := convertMetricsToWindowsStats(metrics)
+	if err != nil {
+		return nil, err
+	}
+
+	podCPU, containerStats, err := c.toPodSandboxStats(sandbox, statsMap, containers, timestamp)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert container stats during podSandboxStats call: %w", err)
+	}
+	podSandboxStats.Windows.Cpu = podCPU.Cpu
+	podSandboxStats.Windows.Memory = podCPU.Memory
+	podSandboxStats.Windows.Containers = containerStats
+
+	podSandboxStats.Windows.Network = windowsNetworkUsage(ctx, sandbox, timestamp)
+
+	pidCount, err := c.getSandboxPidCount(ctx, sandbox)
+	if err != nil {
+		return nil, err
+	}
+
+	podSandboxStats.Windows.Process = &runtime.WindowsProcessUsage{
+		Timestamp:    timestamp.UnixNano(),
+		ProcessCount: &runtime.UInt64Value{Value: pidCount},
+	}
+
+	c.saveSandBoxMetrics(podSandboxStats.Attributes.Id, podSandboxStats)
+
+	return podSandboxStats, nil
+}
+
+func convertMetricsToWindowsStats(metrics []*types.Metric) (map[string]*wstats.Statistics, error) {
+	statsMap := make(map[string]*wstats.Statistics)
+	for _, stat := range metrics {
+		var podRuntimeStats *wstats.Statistics
+		podStatsData, err := typeurl.UnmarshalAny(stat.Data)
+		if err != nil {
+			return nil, fmt.Errorf("failed to extract container metrics: %w", err)
+		}
+
+		// extract the metrics if available
+		// podStatsData can be nil for pods that don't have an actual podsandbox
+		// this is the case for hostprocess pods
+		if podStats, ok := podStatsData.(*wstats.Statistics); ok {
```

Comment:
Suggest we log again if the type assertion fails. e.g. if the stats type changed, would be good to see that in logs as an error instead of silently failing.

Reply:
I guess we should either throw errors in both failure cases in this function, or log in both, depending on if the overall pod stats operation should continue.

Reply:
logging this makes sense to catch the off chance that it changed, but since it is an expected in the case of HPC this isn't a hard error. The previous failure is a hard failure so we should end early.

Reply:
Based on @dcantah feedback, I've wired through the config to tell if this is a HPC, so I can make this a hard error.